presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,703 @@
+    {
+        try {
+            requireNonNull(taskId, "taskId is null");
+            this.taskInstanceId = requireNonNull(taskInstanceId, "taskInstanceIs is null");
+            this.outputBuffers = requireNonNull(outputBuffers, "outputBuffers is null");
+            this.state = requireNonNull(state, "state is null");
+            this.tempStorage = requireNonNull(tempStorage, "tempStorage is null");
+            this.threshold = requireNonNull(threshold, "threshold is null");
+
+            initializeTempStorage();
+            context = new TempDataOperationContext(Optional.empty(), taskId.getQueryId().toString(), Optional.empty(), new Identity("spooling-buffer", Optional.empty()));
+
+            ExecutorService coreExecutor = newCachedThreadPool(daemonThreadsNamed("dispatcher-query-%s"));
+            closer.register(coreExecutor::shutdownNow);
+            executor = listeningDecorator(coreExecutor);
+        }
+        catch (Exception e) {
+            throw new IllegalArgumentException();
+        }
+    }
+
+    @Override
+    public OutputBufferInfo getInfo()
+    {
+        return new OutputBufferInfo(
+                "SPOOLING",
+                state.get(),
+                state.get().canAddBuffers(),
+                state.get().canAddPages(),
+                totalBufferedBytes.get(),
+                totalBufferedPages.get(),
+                totalRowsAdded.get(),
+                totalPagesAdded.get(),
+                ImmutableList.of()
+        );
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return state.get() == BufferState.FINISHED;
+    }
+
+    @Override
+    public double getUtilization()
+    {
+        return 0;
+    }
+
+    @Override
+    public boolean isOverutilized()
+    {
+        return false;
+    }
+
+    @Override
+    public void addStateChangeListener(StateMachine.StateChangeListener<BufferState> stateChangeListener)
+    {
+        state.addStateChangeListener(stateChangeListener);
+    }
+
+    @Override
+    public void setOutputBuffers(OutputBuffers newOutputBuffers)
+    {
+        requireNonNull(newOutputBuffers, "newOutputBuffers is null");
+        checkArgument(outputBuffers.getType() == OutputBuffers.BufferType.SPOOLING, "Invalid output buffers type");
+        checkArgument(outputBuffers.isNoMoreBufferIds(), "invalid noMoreBufferIds");
+
+        if (state.get().isTerminal() || outputBuffers.getVersion() >= newOutputBuffers.getVersion()) {
+            return;
+        }
+
+        outputBuffers.checkValidTransition(newOutputBuffers);
+    }
+
+    @Override
+    public ListenableFuture<BufferResult> get(OutputBuffers.OutputBufferId bufferId, long token, DataSize maxSize)
+    {
+        requireNonNull(bufferId, "outputBufferId is null");
+        checkArgument(bufferId.getId() == 0, "Invalid buffer id");
+        checkArgument(maxSize.toBytes() > 0, "maxSize must be at least 1 byte");
+
+        acknowledge(bufferId, token);
```

Comment:
Because this method mutates `handleInfoQueue` and `pages`, we'll want to do this inside the synchronized block.  This coud complicate things because there's some blocking IO calls there (`TempStorageHandle.remove`).  As a solution, we can perform the cleanup in a fire-and-forget executor (whatever temp storage we use, there is always some sort of backup GC mechanism to remove stray files).